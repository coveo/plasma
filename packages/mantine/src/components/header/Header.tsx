import {QuestionSize16Px} from '@coveord/plasma-react-icons';
import {Anchor, Breadcrumbs, DefaultProps, Divider, Group, Stack, Text, Title, Tooltip} from '@mantine/core';
import {Children, FunctionComponent, ReactElement, ReactNode} from 'react';

export interface HeaderProps extends DefaultProps {
    /**
     * The description text displayed inside the header underneath the title
     */
    description?: ReactNode;
    /**
     * Whether the header should have a border on the bottom
     */
    borderBottom?: boolean;
    /**
     * Use the modal variant when displaying the header inside a modal
     *
     * @default 'page'
     */
    variant?: 'page' | 'modal';
    /**
     * The title of the header.
     */
    children: ReactNode;
}

interface HeaderType {
    (props: HeaderProps): ReactElement;
    Breadcrumbs: typeof HeaderBreadcrumbs;
    Actions: typeof HeaderActions;
    DocAnchor: typeof HeaderDocAnchor;
}

export const Header: HeaderType = ({description, borderBottom, children, variant = 'page', ...others}) => {
    const convertedChildren = Children.toArray(children) as ReactElement[];
    const breadcrumbs = convertedChildren.find((child) => child.type === HeaderBreadcrumbs);
    const actions = convertedChildren.find((child) => child.type === HeaderActions);
    const docAnchor = convertedChildren.find((child) => child.type === HeaderDocAnchor);
    const otherChildren = convertedChildren.filter(
        (child) => child.type !== HeaderBreadcrumbs && child.type !== HeaderActions && child.type !== HeaderDocAnchor,
    );
    return (
        <>
            <Group
                position="apart"
                p={variant === 'page' ? 'xl' : undefined}
                pb={variant === 'page' ? 'lg' : undefined}
                {...others}
            >
                <Stack spacing={0}>
                    {breadcrumbs}
                    <Title
                        order={variant === 'page' ? 1 : 3}
                        color={variant === 'page' ? 'gray.5' : undefined}
                        sx={{wordBreak: 'break-word'}}
                    >
                        {otherChildren}
                        {docAnchor}
                    </Title>
                    <Text size={variant === 'page' ? 'md' : 'sm'} color="gray.6">
                        {description}
                    </Text>
                </Stack>
                {actions}
            </Group>
            {borderBottom ? <Divider size="xs" /> : null}
        </>
    );
};

const HeaderBreadcrumbs: FunctionComponent<{children: ReactNode}> = ({children}) => (
    <Breadcrumbs
        styles={(theme) => ({
            breadcrumb: {fontSize: theme.fontSizes.sm, fontWeight: 300},
            separator: {color: theme.colors.gray[5]},
        })}
    >
        {children}
    </Breadcrumbs>
);

const HeaderActions: FunctionComponent<{children: ReactNode}> = ({children}) => <Group spacing="sm">{children}</Group>;

export interface HeaderDocAnchorProps {
    /**
     * A href pointing to documentation related to the current panel.
     * When provided, an info icon is rendered next to the title as link to this documentation
     */
    href: string;
    /**
     * The tooltip text shown when hovering over the doc link icon
     */
    label?: string;
}

const HeaderDocAnchor: FunctionComponent<HeaderDocAnchorProps> = ({href: docLink, label: docLinkTooltipLabel}) => (
    <Tooltip label={docLinkTooltipLabel} disabled={!docLinkTooltipLabel} position="right">
        <Anchor inline href={docLink} target="_blank" ml="xs" style={{verticalAlign: 'middle'}}>
            <QuestionSize16Px height={16} />
        </Anchor>
    </Tooltip>
);

Header.Breadcrumbs = HeaderBreadcrumbs;
Header.Actions = HeaderActions;
Header.DocAnchor = HeaderDocAnchor;
