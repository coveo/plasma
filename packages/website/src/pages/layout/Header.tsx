import {HeaderMetadata} from '@coveord/plasma-components-props-analyzer';
import mainExample from '@examples/layout/Header/Header.example.tsx';

import {PageLayout} from '../../building-blocs/PageLayout';

export default () => (
    <PageLayout
        section="Layout"
        title="Header"
        sourcePath="/packages/mantine/src/components/header/Header.tsx"
        description="A page header informs a user of the section of the product they are currently in. It includes a breadcrumb and optional call to actions."
        thumbnail="header"
        id="Header"
        propsMetadata={HeaderMetadata}
        code={mainExample}
    />
);
