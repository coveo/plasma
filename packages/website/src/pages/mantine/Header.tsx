import {HeaderMantineMetadata} from '@coveord/plasma-components-props-analyzer';
import mainExample from '@examples/Header/Header.mantine.example.tsx';

import {PageLayout} from '../../building-blocs/PageLayout';

export default () => (
    <PageLayout
        section="Mantine"
        title="Page Header"
        sourcePath="/packages/mantine/src/components/header/Header.tsx"
        description="A page header informs a user of the section of the product they are currently in. It includes a breadcrumb and optional tabs."
        thumbnail="header"
        id="Header"
        propsMetadata={HeaderMantineMetadata}
        code={mainExample}
    />
);
