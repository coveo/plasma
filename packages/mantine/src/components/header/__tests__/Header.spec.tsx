import {Anchor} from '@mantine/core';
import {render, screen} from '@test-utils';

import {Header} from '../Header';

describe('Header', () => {
    it('renders a heading of level 1 with the specified children within', () => {
        render(<Header>child</Header>);

        const header = screen.getByRole('heading');
        expect(header).toMatchInlineSnapshot(`
          <h1
            class="mantine-Text-root mantine-Title-root mantine-PlasmaHeader-title mantine-1o7jx33"
          >
            child
          </h1>
        `);
    });

    it('renders the specified breadcrumbs above the title', async () => {
        const {container, rerender} = render(
            <Header>
                Title
                <Header.Breadcrumbs>
                    <Anchor>One</Anchor>
                    <Anchor>Two</Anchor>
                    <Anchor>Three</Anchor>
                </Header.Breadcrumbs>
            </Header>,
        );
        const titleFirst = container;

        rerender(
            <Header>
                <Header.Breadcrumbs>
                    <Anchor>One</Anchor>
                    <Anchor>Two</Anchor>
                    <Anchor>Three</Anchor>
                </Header.Breadcrumbs>
                Title
            </Header>,
        );
        const breadcrumbsFirst = container;

        expect(titleFirst).toMatchSnapshot();
        expect(breadcrumbsFirst).toEqual(titleFirst);
    });

    it('renders a doc link icon if a doc anchor is provided', async () => {
        render(
            <Header>
                title
                <Header.DocAnchor href="/some/path" />
            </Header>,
        );

        const docLink = await screen.findByRole('link', {name: /question/i});
        expect(docLink).toHaveAttribute('href', '/some/path');
    });

    it('adds a separator underneath the header if borderBottom is true', () => {
        render(<Header borderBottom>title</Header>);
        expect(screen.getByRole('separator')).toBeInTheDocument();
    });

    it('renders provided description in the header', () => {
        render(<Header description="description">title</Header>);

        expect(screen.getByText('description')).toBeInTheDocument();
    });

    it('renders provided actions in the header', () => {
        render(
            <Header>
                title
                <Header.Actions>
                    <span>action 1</span>
                    <span>action 2</span>
                </Header.Actions>
            </Header>,
        );

        expect(screen.getByText('action 1')).toBeInTheDocument();
        expect(screen.getByText('action 2')).toBeInTheDocument();
    });
});
