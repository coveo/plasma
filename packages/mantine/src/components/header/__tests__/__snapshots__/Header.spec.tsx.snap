// Vitest Snapshot v1

exports[`Header > renders the specified breadcrumbs above the title 1`] = `
<div>
  <div
    class="mantine-Group-root mantine-ogmezo"
  >
    <div
      class="mantine-Stack-root mantine-1178y6y"
    >
      <div
        class="mantine-Breadcrumbs-root mantine-1ujbd2v"
      >
        <a
          class="mantine-Text-root mantine-Anchor-root mantine-Breadcrumbs-breadcrumb mantine-1yycs9g"
        >
          One
        </a>
        <div
          class="mantine-Text-root mantine-Breadcrumbs-separator mantine-c5usyo"
        >
          /
        </div>
        <a
          class="mantine-Text-root mantine-Anchor-root mantine-Breadcrumbs-breadcrumb mantine-1yycs9g"
        >
          Two
        </a>
        <div
          class="mantine-Text-root mantine-Breadcrumbs-separator mantine-c5usyo"
        >
          /
        </div>
        <a
          class="mantine-Text-root mantine-Anchor-root mantine-Breadcrumbs-breadcrumb mantine-1yycs9g"
        >
          Three
        </a>
      </div>
      <div
        class="mantine-xg7kom"
      >
        <h1
          class="mantine-Text-root mantine-Title-root mantine-d1yoif"
        >
          Title
        </h1>
      </div>
      <div
        class="mantine-Text-root mantine-1w25z6f"
      />
    </div>
  </div>
</div>
`;
